contracts/minter-imago/src/contract.rs

```diff
@@ -0,0 +1,743 @@
+        return Err(ContractError::MaxPerAddressLimitExceeded {});
+    }
+
+    Ok(false)
+}
+
+pub fn execute_mint_to(
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    recipient: String,
+) -> Result<Response, ContractError> {
+    let recipient = deps.api.addr_validate(&recipient)?;
+    let config = CONFIG.load(deps.storage)?;
+    let action = "mint_to";
+
+    // Check only admin
+    if info.sender != config.admin {
+        return Err(ContractError::Unauthorized(
+            "Sender is not an admin".to_owned(),
+        ));
+    }
+
+    _execute_mint(deps, env, info, action, true, Some(recipient), None)
+}
+
+pub fn execute_mint_for(
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    token_id: u32,
+    recipient: String,
+) -> Result<Response, ContractError> {
+    let recipient = deps.api.addr_validate(&recipient)?;
+    let config = CONFIG.load(deps.storage)?;
+    let action = "mint_for";
+
+    // Check only admin
+    if info.sender != config.admin {
+        return Err(ContractError::Unauthorized(
+            "Sender is not an admin".to_owned(),
+        ));
+    }
+
+    _execute_mint(deps, env, info, action, true, Some(recipient), Some(token_id))
+}
+
+// Generalize checks and mint message creation
+// mint -> _execute_mint(recipient: None, token_id: None)
+// mint_to(recipient: "friend") -> _execute_mint(Some(recipient), token_id: None)
+// mint_for(recipient: "friend2", token_id: 420) -> _execute_mint(recipient, token_id)
+fn _execute_mint(
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    action: &str,
+    is_admin: bool,
+    recipient: Option<Addr>,
+    token_id: Option<u32>,
+) -> Result<Response, ContractError> {
+    let config = CONFIG.load(deps.storage)?;
+    let sg721_address = SG721_ADDRESS.load(deps.storage)?;
+
+    let recipient_addr = match recipient {
+        Some(some_recipient) => some_recipient,
+        None => info.sender.clone(),
+    };
+
+    let mint_price: Coin = mint_price(deps.as_ref(), is_admin)?;
+    // Exact payment only accepted
+    let payment = may_pay(&info, &config.unit_price.denom)?;
+    if payment != mint_price.amount {
+        return Err(ContractError::IncorrectPaymentAmount(
+            coin(payment.u128(), &config.unit_price.denom),
+            mint_price,
+        ));
+    }
+
+    let mut msgs: Vec<CosmosMsg<StargazeMsgWrapper>> = vec![];
+
+    // Create network fee msgs
+    let fee_percent = if is_admin {
+        Decimal::percent(AIRDROP_MINT_FEE_PERCENT as u64)
+    } else {
+        Decimal::percent(MINT_FEE_PERCENT as u64)
+    };
+    let network_fee = mint_price.amount * fee_percent;
+    let pw_fee = mint_price.amount * Decimal::percent(PW_MINT_FEE_PERCENT);
+    let addr = maybe_addr(deps.api, Some(DEV_ADDRESS.to_string()))?;
+    msgs.append(&mut pw_fee_msg(&info, pw_fee.u128(), addr.clone().unwrap())?);
+    msgs.append(&mut checked_fair_burn(&info, network_fee.u128(), addr.clone())?);
```

Comment:
this will be taking 2 fees there is a 10% for stargaze + 15% fee pw right ?

Reply:
Yes, I made the assumption that my contract should maintain all the stargaze fees. Let me know if you have other thoughts on this.

Reply:
We have dev fee in `SG1 Fairburn` https://github.com/public-awesome/launchpad/blob/main/packages/sg1/src/lib.rs#L32, might be easier to use that since fair burn % and dev % could change w governance proposals

Reply:
I am using dev fee. It's not a configurable fee so I'm also taking an additional 15% mint fee to cover public works costs.